types/Config.js

```diff
@@ -210,6 +210,7 @@ export type ProjectConfig = {|
   cache: boolean,
   cacheDirectory: Path,
   clearMocks: boolean,
+  console?: Object,
```

Comment:
can you just do `console?: console`? Flow has it typed out: https://github.com/facebook/flow/blob/e2325d0e216e0cc14d17f00fc96624b364a27902/lib/core.js#L812-L835

Reply:
Hmmm... When I try that I get:

> object type. Ineligible value used in/as type annotation (did you forget 'typeof'?)

I'm still pretty new to flow so maybe I'm doing something wrong... I did:

```javascript
console?: console,
```

Is that wrong?

Reply:
I'm really bad at flow myself :P Try the suggestion of `console?: typeof console`

Reply:
Haha, that gives me a crazy amount of errors. Most of them look unrelated, but some are:

```
Error: packages/jest-runner/src/run_test.js:97
 97:     Object.assign({}, config, {console: testConsole}),
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ object literal. This type is incompatible with the expected param type of
 16:   constructor(config: ProjectConfig): void;
                           ^^^^^^^^^^^^^ object type. See: types/Environment.js:16
  Property `console` is incompatible:
     97:     Object.assign({}, config, {console: testConsole}),
                                                 ^^^^^^^^^^^ BufferedConsole. This type is incompatible with
    213:   console?: typeof console,
                     ^^^^^^^^^^^^^^ object type. See: types/Config.js:213
      Property `dirxml` is incompatible:
        213:   console?: typeof console,
                         ^^^^^^^^^^^^^^ property `dirxml`. Property not found in. See: types/Config.js:213
         97:     Object.assign({}, config, {console: testConsole}),
                                                     ^^^^^^^^^^^ BufferedConsole
```

🤷‍♂️

Reply:
Heh, yeah that's a bit too strict I guess. Thanks for checking!

Reply:
Hey, just circling back on this again. I will not be able to merge this diff unless we stop adding this field to the ProjectConfig type. First, all data in an object conforming to this type has to be JSON serializable because it gets sent over worker processes. Second, there is no point in polluting this object type to pass additional data to one consumer of this type only. I'm not arguing here about the functionality, I'm arguing about muddling what this type is used for.

Reply:
I'm fine with that 👌 

FWIW, the `console` object is actually only added to a clone of the `config` object so it wont ever be sent to be serialized for a worker, it's only used when initializing the worker.

That said I understand what you mean. What would you suggest we do then? Create another argument for the TestEnvironment constructor? I'm happy to do that :+1:

Reply:
Yap, whether adding this property when it's passed over a worker or not doesn't really change the purpose of the type though. Another solution would be to use a union type but I definitely would prefer adding the console prop as an argument to the TestEnvironment constructor. Since we still have the `setGlobal` call, this should be backward compatible.

Reply:
Sounds good. Do you think I should make it an object that has a `console` property? That would provide the flexibility to adding more properties if other test environments need something like this in the future.

Reply:
anybody extending JSDOMEnvironment would then not get this improvement for free. Might be edge-casey enough to ignore, I guess.

<sup>Could add a `// $FlowIgnoreMe` :D</sup>

Reply:
Yeah, I think passing the "config" object together with an "options" object is my preferred solution here.

Reply:
Sounds good. I'll do that tomorrow then. It's 1:30 AM where I am :sweat_smile: